File: src/java/pulse/cdm/patient/actions/SEForcedInhale.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.patient.actions;

import pulse.cdm.bind.PatientActions.ForcedInhaleData;

import pulse.cdm.properties.SEScalar0To1;
import pulse.cdm.properties.SEScalarTime;

public class SEForcedInhale extends SEConsciousRespirationCommand
{
  protected SEScalar0To1 inspiratoryCapacityFraction;
  protected SEScalarTime inhalePeriod;
  protected SEScalarTime holdPeriod;
  protected SEScalarTime releasePeriod;
  
  public SEForcedInhale()
  {
    inspiratoryCapacityFraction = null;
    inhalePeriod = null;
    holdPeriod = null;
    releasePeriod = null;
  }
  
  public void copy(SEForcedInhale other)
  {
    if(this==other)
      return;
    
    if(other.hasInspiratoryCapacityFraction())
      getInspiratoryCapacityFraction().set(other.inspiratoryCapacityFraction);
    if(other.hasInspiratoryCapacityFraction())
      getInspiratoryCapacityFraction().set(other.inspiratoryCapacityFraction);
    if(other.hasInspiratoryCapacityFraction())
      getInspiratoryCapacityFraction().set(other.inspiratoryCapacityFraction);
    if(other.hasInspiratoryCapacityFraction())
      getInspiratoryCapacityFraction().set(other.inspiratoryCapacityFraction);
  }
  
  public SEConsciousRespirationCommand getCopy()
  {
    SEConsciousRespirationCommand command = new SEForcedInhale();
    command.copy(this);
    
    return command;
  }
  
  public void reset()
  {
    super.reset();
    if (inspiratoryCapacityFraction != null)
      inspiratoryCapacityFraction.invalidate();
    if (inhalePeriod != null)
      inhalePeriod.invalidate();
    if (holdPeriod != null)
      holdPeriod.invalidate();
    if (releasePeriod != null)
      releasePeriod.invalidate();
  }
  
  public boolean isValid()
  {
    return hasInspiratoryCapacityFraction() && hasInhalePeriod();
  }
  
  public static void load(ForcedInhaleData src, SEForcedInhale dst)
  {
    //SEConsciousRespirationCommand.load(src.getConsciousRespirationCommand(), dst);
    if(src.hasInspiratoryCapacityFraction())
      SEScalar0To1.load(src.getInspiratoryCapacityFraction(),dst.getInspiratoryCapacityFraction());
    if(src.hasInhalePeriod())
      SEScalarTime.load(src.getInhalePeriod(),dst.getInhalePeriod());
    if(src.hasHoldPeriod())
      SEScalarTime.load(src.getHoldPeriod(),dst.getHoldPeriod());
    if(src.hasReleasePeriod())
      SEScalarTime.load(src.getReleasePeriod(),dst.getReleasePeriod());
  }
  
  public static ForcedInhaleData unload(SEForcedInhale src)
  {
    ForcedInhaleData.Builder dst = ForcedInhaleData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SEForcedInhale src, ForcedInhaleData.Builder dst)
  {
    //SEConsciousRespirationCommand.unload(src,dst.getConsciousRespirationCommand());    
    if (src.hasInspiratoryCapacityFraction())
      dst.setInspiratoryCapacityFraction(SEScalar0To1.unload(src.inspiratoryCapacityFraction));
    if (src.hasInhalePeriod())
      dst.setInhalePeriod(SEScalarTime.unload(src.inhalePeriod));
    if (src.hasHoldPeriod())
      dst.setHoldPeriod(SEScalarTime.unload(src.holdPeriod));
    if (src.hasReleasePeriod())
      dst.setReleasePeriod(SEScalarTime.unload(src.releasePeriod));
  }
  
  public boolean hasInspiratoryCapacityFraction()
  {
    return inspiratoryCapacityFraction != null;
  }
  public SEScalar0To1 getInspiratoryCapacityFraction()
  {
    if (inspiratoryCapacityFraction == null)
      inspiratoryCapacityFraction = new SEScalar0To1();
    return inspiratoryCapacityFraction;
  }
  
  public boolean hasInhalePeriod()
  {
    return inhalePeriod != null;
  }
  public SEScalarTime getInhalePeriod()
  {
    if (inhalePeriod == null)
      inhalePeriod = new SEScalarTime();
    return inhalePeriod;
  }
  
  public boolean hasHoldPeriod()
  {
    return holdPeriod != null;
  }
  public SEScalarTime getHoldPeriod()
  {
    if (holdPeriod == null)
      holdPeriod = new SEScalarTime();
    return holdPeriod;
  }
  
  public boolean hasReleasePeriod()
  {
    return releasePeriod != null;
  }
  public SEScalarTime getReleasePeriod()
  {
    if (releasePeriod == null)
      releasePeriod = new SEScalarTime();
    return releasePeriod;
  }
  
  public String toString()
  {
      return "Forced Inhale" 
          + "\n\tInspiratoryCapacityFraction: " + (hasInspiratoryCapacityFraction() ? getInspiratoryCapacityFraction() : "Not Specified")
          + "\n\tInhalePeriod: " + (hasInhalePeriod() ? getInhalePeriod() : "Not Specified")
          + "\n\tHoldPeriod: " + (hasHoldPeriod() ? getHoldPeriod() : "Not Specified")
          + "\n\tReleasePeriod: " + (hasReleasePeriod() ? getReleasePeriod() : "Not Specified");
  }
}

File: src/cpp/cdm/patient/actions/SEForcedInhale.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SEForcedInhale.h"
#include "properties/SEScalar0To1.h"
#include "properties/SEScalarTime.h"
#include "io/protobuf/PBPatientActions.h"

SEForcedInhale::SEForcedInhale(Logger* logger) : SEConsciousRespirationCommand(logger)
{
  m_InspiratoryCapacityFraction = nullptr;
  m_InhalePeriod = nullptr;
  m_HoldPeriod = nullptr;
  m_ReleasePeriod = nullptr;
}

SEForcedInhale::~SEForcedInhale()
{
  SAFE_DELETE(m_InspiratoryCapacityFraction);
  SAFE_DELETE(m_InhalePeriod);
  SAFE_DELETE(m_HoldPeriod);
  SAFE_DELETE(m_ReleasePeriod);
}

void SEForcedInhale::Clear()
{
  SEConsciousRespirationCommand::Clear();
  INVALIDATE_PROPERTY(m_InspiratoryCapacityFraction);
  INVALIDATE_PROPERTY(m_InhalePeriod);
  INVALIDATE_PROPERTY(m_HoldPeriod);
  INVALIDATE_PROPERTY(m_ReleasePeriod);
}

void SEForcedInhale::Copy(const SEForcedInhale& src)
{
  PBPatientAction::Copy(src, *this);
}

bool SEForcedInhale::IsValid() const
{
  return SEConsciousRespirationCommand::IsValid() && HasInspiratoryCapacityFraction() && HasInhalePeriod();
}

bool SEForcedInhale::IsActive() const
{
  return SEConsciousRespirationCommand::IsActive();
}

bool SEForcedInhale::HasInspiratoryCapacityFraction() const
{
  return m_InspiratoryCapacityFraction == nullptr ? false : m_InspiratoryCapacityFraction->IsValid();
}
SEScalar0To1& SEForcedInhale::GetInspiratoryCapacityFraction()
{
  if (m_InspiratoryCapacityFraction == nullptr)
    m_InspiratoryCapacityFraction = new SEScalar0To1();
  return *m_InspiratoryCapacityFraction;
}
double SEForcedInhale::GetInspiratoryCapacityFraction() const
{
  if (m_InspiratoryCapacityFraction == nullptr)
    return SEScalar::dNaN();
  return m_InspiratoryCapacityFraction->GetValue();
}

bool SEForcedInhale::HasInhalePeriod() const
{
  return m_InhalePeriod == nullptr ? false : m_InhalePeriod->IsValid();
}
SEScalarTime& SEForcedInhale::GetInhalePeriod()
{
  if (m_InhalePeriod == nullptr)
    m_InhalePeriod = new SEScalarTime();
  return *m_InhalePeriod;
}
double SEForcedInhale::GetInhalePeriod(const TimeUnit& unit) const
{
  if (m_InhalePeriod == nullptr)
    return SEScalar::dNaN();
  return m_InhalePeriod->GetValue(unit);
}

bool SEForcedInhale::HasHoldPeriod() const
{
  return m_HoldPeriod == nullptr ? false : m_HoldPeriod->IsValid();
}
SEScalarTime& SEForcedInhale::GetHoldPeriod()
{
  if (m_HoldPeriod == nullptr)
    m_HoldPeriod = new SEScalarTime();
  return *m_HoldPeriod;
}
double SEForcedInhale::GetHoldPeriod(const TimeUnit& unit) const
{
  if (m_HoldPeriod == nullptr)
    return SEScalar::dNaN();
  return m_HoldPeriod->GetValue(unit);
}

bool SEForcedInhale::HasReleasePeriod() const
{
  return m_ReleasePeriod == nullptr ? false : m_ReleasePeriod->IsValid();
}
SEScalarTime& SEForcedInhale::GetReleasePeriod()
{
  if (m_ReleasePeriod == nullptr)
    m_ReleasePeriod = new SEScalarTime();
  return *m_ReleasePeriod;
}
double SEForcedInhale::GetReleasePeriod(const TimeUnit& unit) const
{
  if (m_ReleasePeriod == nullptr)
    return SEScalar::dNaN();
  return m_ReleasePeriod->GetValue(unit);
}

void SEForcedInhale::ToString(std::ostream &str) const
{
  str << "Forced Inhale";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  str << "\n\tInspiratoryCapacityFraction: "; HasInspiratoryCapacityFraction() ? str << *m_InspiratoryCapacityFraction : str << "NaN";
  str << "\n\tInhalePeriod: "; HasInhalePeriod() ? str << *m_InhalePeriod : str << "NaN";
  str << "\n\tHoldPeriod: "; HasHoldPeriod() ? str << *m_HoldPeriod : str << "NaN";
  str << "\n\tReleasePeriod: "; HasReleasePeriod() ? str << *m_ReleasePeriod : str << "NaN";
  str << std::flush;
}